students/kristianjf/lesson09/mailroom_5.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python3
+import sys
+'''
+Goal:
+You work in the mail room at a local charity. Part of your job is to write \
+incredibly boring, repetitive emails thanking your donors for their generous \
+gifts. You are tired of doing this over and over again, so you’ve decided to \
+let Python help you out of a jam and do your work for you.
+'''
+'''
+It should have a data structure that holds a list of your donors and a history of \
+the amounts they have donated. This structure should be populated at first with at \
+least five donors, with between 1 and 3 donations each.
+'''
+# self.DATA_STRUCTURE = [('Jeff Bezos', [200, 500, 1000]), \
+#                     ('Bill Gates', [1000]), \
+#                     ('Steve Jobs', [20, 50, 100])]
+# self.DATA_STRUCTURE_DICTS = [{'name': name, 'donation': donation} for name, donation in self.DATA_STRUCTURE]
+# # self.DATA_STRUCTURE_DICT = [ \
+# #                         {'name': 'Jeff Bezos', 'donation': [200, 500, 1000]},\
+# #                         {'name': 'Bill Gates', 'donation': [1000]},\
+# #                         {'name': 'Steve Jobs', 'donation': [20, 50, 100]}\
+# #                         ]
+
+class Donor:
+    def __init__(self, name, donations):
+        self._name = name
+        self._donations = [] if donations is None else donations
+    @property
+    def name(self):
+        return self._name
+    @property
+    def donations(self):
+        return self._donations
+    def donate(self, new_donation):
+        self.donations.append(new_donation)
+    def metrics(self):
+        self.num_d = len(self.donations)
+        self.total_d = sum(self.donations)
+        self.avg_d = self.total_d / self.num_d
+
+class DonorsHandle:
+    DATA_STRUCTURE = [] #[('',[])]
```

Comment:
putting these as class attributes means that each instance of the class will have the same copies.

You typically only wil have one instance, so that's not too bad, but if you do make ore than one, it could get ugly. If nothign else, creating instances for tests is helpful.

So better to make these instance attributes, but defining them in __init__:

```
def __init__(self):
    self.data_structure = []
    self.data_structure_dicts = []
```
Though a bit odd to have both of those -- but maybe I'll see why as I look at more of the code :-)